### PR TITLE
Adding Python Gen AI pipelines implementation

### PR DESCRIPTION
**Pull request by riacheruvu**
The purpose of this PR is a proposal around refactoring the existing Python implementation in the notebooks for Gen AI models (e.g., SD variations, LCM, etc.) and add them to the repository, complementing the structure of the C++ pipelines.

Please note: This is a draft PR. If this initial contribution makes sense, I can also add in the txt2img pipeline refactoring from the notebook, the quantization processes as part of the documentation README, Python implementations for llama2, MPT, and the other img gen models such as Würstchen, etc. Thanks!

@yury-gorbachev

--
Update 12/8/2023:

Added models from notebooks:

- Audio Generation: Distil Whisper
- Image Generation (Stable Diffusion): Stable Diffusion v2.1, TinySD, SD v2.5
- Image Generation (Stable Diffusion XL): SDXL, SSD B1

WIP models: SDXL Turbo, LLM

### CODE REVIEW

**Comment by slyalin** [comment]
@riacheruvu, I haven't yet dig into the code itself, but there is rather technical comment: please do not bring `.ipynb_checkpoints` and `__pycache__` directories to the PR.

**Comment by riacheruvu** [comment]
> @riacheruvu, I haven't yet looked into the code itself, but there is rather technical comment: please do not bring `.ipynb_checkpoints` and `__pycache__` directories to the PR.

Fixed, thank you @slyalin. My apologies for the oversight.

File: image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
+from typing import Union, Optional, Any, List, Dict
+from transformers import CLIPTokenizer, CLIPImageProcessor
+from diffusers.pipelines.stable_diffusion.safety_checker import (
+    StableDiffusionSafetyChecker,
+)
+from diffusers.pipelines.stable_diffusion import StableDiffusionPipelineOutput
+from diffusers.image_processor import VaeImageProcessor
+from diffusers import DiffusionPipeline
```
Comment (slyalin) [inline image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py:8]: I think the purpose of Python examples in this repository is to demonstrate how to work with the models without extra dependencies in almost native Python+openvino environment. Here we are importing some stuff with quite heavy dependencies like `torch`. It should be only applicable for code that does export of the models, but not the inference code.

Comment (riacheruvu) [reply image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py:8]: Noted, thank you! I'm reviewing the unnecessary dependencies and will take these out in the next commit.

File: image_generation/stable_diffusion_xl/main.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+core = Core()
+device = "GPU"
+text_prompt = "valley in the Alps at sunset, epic vista, beautiful landscape, 4k, 8k"
+neg_prompt = None
+#neg_prompt = "ugly, blurry, poor quality" #Negative prompt for SSD-B1 model
+num_inference_steps = 25
+seed = 42
+
+#Define and load the model
+text2image_pipe = OVStableDiffusionXLPipeline.from_pretrained(model_dir, device=device)
```
Comment (riacheruvu) [inline image_generation/stable_diffusion_xl/main.py:21]: @slyalin, is one of the goals to directly including the Optimum-Intel pipeline implementations in the repo (e.g., [OVStableDiffusionXLPipeline ](https://github.com/huggingface/optimum-intel/blob/f32d5015de83ec561e8e4b7d3b19fdb8ecd6595f/optimum/intel/openvino/modeling_diffusion.py#L833) in the repo? Thank you!

File: image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py (hunk 1 above)
Comment (riacheruvu) [reply image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py]: @slyalin, I had two follow-up questions please: 

1. I removed the CLIPTokenizer and Image Processor dependencies. The other dependencies are being used in the inference pipeline. Would it make sense to export the VaeImageProcessor model offline, since I notice this is an API for input prep?

3. I see the tokenizer export is in the llm benchmarking part of the repo - is it useful to replace LCM's original tokenizer with the OVtokenizer, or is this not the right time? Thank you!
```
#Define LCM scheduler, tokenizer
#TBD - access pre-saved scheduler and tokenizer
pipe = DiffusionPipeline.from_pretrained("SimianLuo/LCM_Dreamshaper_v7")
scheduler = pipe.scheduler
tokenizer = pipe.tokenizer
```

File: image_generation/stable_diffusion_xl/main.py (hunk 2 above)
Comment (ilya-lavrenov) [reply image_generation/stable_diffusion_xl/main.py:21]: the goal is to avoid dependency on HF and implement missing parts (scheduler in this case) from scratch

Comment (riacheruvu) [reply image_generation/stable_diffusion_xl/main.py:21]: Thank you, @ilya-lavrenov! I can look into porting the [LMSDiscreteScheduler ](https://github.com/openvinotoolkit/openvino.genai/blob/4b0a2edd709a61b85cc407be2ea79de594dc5925/image_generation/stable_diffusion_1_5/cpp/src/stable_diffusion.hpp#L75C20-L75C40) to Python.

I noticed in another places of the repo, [the scheduler is also saved](https://github.com/openvinotoolkit/openvino.genai/blob/4b0a2edd709a61b85cc407be2ea79de594dc5925/llm_bench/python/convert.py#L358) - is this another approach to consider?

File: audio_generation/distil-whisper/main.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from openvino.runtime import Core
+import torch
+from datasets import load_dataset
+from transformers import AutoProcessor
+from optimum.intel.openvino import OVModelForSpeechSeq2Seq
+
+#Define model parameters
+distil_model_id = "distil-whisper/distil-large-v2"
+core = Core()
+device = "GPU"
```
Comment (AlexKoff88) [inline audio_generation/distil-whisper/main.py:11]: I would make"CPU" as a default device since GPU is not always available.

File: audio_generation/distil-whisper/main.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+processor = AutoProcessor.from_pretrained(distil_model_id)
+
+#Specify model dir
+model_dir = Path(distil_model_id.split("/")[-1])
+
+#Prepare and load dataset
+def extract_input_features(sample):
```
Comment (AlexKoff88) [inline audio_generation/distil-whisper/main.py:18]: This example is fully based on Optimum-Intel. @riacheruvu, don't you think that it should be a part of this [documentation](https://huggingface.co/docs/optimum/intel/inference)?

File: image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+import openvino as ov
+import torch
+
+class OVLatentConsistencyModelPipeline(DiffusionPipeline):
```
Comment (AlexKoff88) [inline image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py:11]: I think this sample replicates what we have in Optimum-Intel. What is the value of this class then? Please note that both (Optimum version and this one) will require maintenance.

File: image_generation/latent_consistency_model/main.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+from pathlib import Path
+from openvino.runtime import Core
+from diffusers.schedulers import LMSDiscreteScheduler
+from transformers import CLIPTokenizer
+from OVLatentConsistencyModelPipeline import OVLatentConsistencyModelPipeline
+import torch
+from diffusers import DiffusionPipeline
```
Comment (AlexKoff88) [inline image_generation/latent_consistency_model/main.py:7]: The same comment as above is that it should be a part of Optimum [documentation](https://huggingface.co/docs/optimum/intel/inference). I think the main idea of Gen.AI samples is to show examples of deployment using OpenVINO native API with minimum dependencies on 3rd party libraries, e.g. using numpy only.

File: image_generation/stable_diffusion/python/OVStableDiffusionPipeline.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    pad_height = 512 - dst_height
+    pad = ((0, 0), (0, pad_height), (0, pad_width), (0, 0))
+    image = np.pad(image, pad, mode="constant")
+    image = image.astype(np.float32) / 255.0
+    image = 2.0 * image - 1.0
+    image = image.transpose(0, 3, 1, 2)
+    return image, {"padding": pad, "src_width": src_width, "src_height": src_height}
+
+
+class OVStableDiffusionPipeline(DiffusionPipeline):
```
Comment (AlexKoff88) [inline image_generation/stable_diffusion/python/OVStableDiffusionPipeline.py:61]: The same comment as for LCM.

File: image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py (hunk 5 above)
Comment (riacheruvu) [reply image_generation/latent_consistency_model/OVLatentConsistencyModelPipeline.py:11]: @AlexKoff88, thank you for the detailed comments. Right now, the versions I've pushed in the draft PR are identical to the optimum samples - I'm working on isolating the HF dependencies step by step in the pipeline to incorporate the OV native runtime API. Per our chat, I will modify the snippets to show the native workflow of export from optimum -> compile, without instantiating the Pipeline classes. Thank you for the direction!
